File: sdk/node/example/src/regression/run_reconnect_test.ts
import {
    ClientOptionBuilder,
    DefaultClient,
    GlobalApiEndpoint,
    GlobalBrokerApiEndpoint,
    GlobalFuturesApiEndpoint,
    Spot,
    TransportOptionBuilder,
    WebSocketClientOptionBuilder,
} from 'kucoin-universal-sdk';

const SLEEP_SECONDS = 5;

function callbackFunc(_topic: string, _subject: string, _data: any): void {
}

async function getSpotSymbols(restService: any): Promise<string[]> {
    const marketApi = restService.getSpotService().getMarketApi();
    const resp = await marketApi.getAllSymbols(
        Spot.Market.GetAllSymbolsReq.builder().setMarket('USDS').build(),
    );
    const symbols: string[] = resp.data.map((d: any) => d.symbol);
    return symbols.length > 100 ? symbols.slice(0, 100) : symbols;
}

async function spotWsExample(ws: any, symbols: string[]): Promise<void> {
    let p = ws.start();
    for (const s of symbols) {
        p = p.then(() => {
            return ws.trade([s], callbackFunc);
        });
    }
    p = p.then(() => {
        return ws.ticker(['BTC-USDT', 'ETH-USDT'], callbackFunc);
    });

    return p.then(() => {
        console.info('Spot subscribe [OK]');
    }).catch((e: any) => {
        console.error('Spot subscribe [Error]', e);
    });
}

async function futuresWsExample(ws: any): Promise<void> {
    return ws.start().then(() => {
        return ws.tickerV2('XBTUSDTM', callbackFunc);
    }).then(() => {
        return ws.tickerV1('XBTUSDTM', callbackFunc);
    }).then(() => {
        console.info('Futures subscribe [OK]');
    }).catch((e: any) => {
        console.error('Futures subscribe [Error]', e);
    })
}

async function wsReconnectTest(): Promise<void> {
    const key = process.env.API_KEY ?? '';
    const secret = process.env.API_SECRET ?? '';
    const passphrase = process.env.API_PASSPHRASE ?? '';

    const wsOption = new WebSocketClientOptionBuilder().build();

    const clientOption = new ClientOptionBuilder()
        .setKey(key)
        .setSecret(secret)
        .setPassphrase(passphrase)
        .setWebSocketClientOption(wsOption)
        .setTransportOption(new TransportOptionBuilder().build())
        .setSpotEndpoint(GlobalApiEndpoint)
        .setFuturesEndpoint(GlobalFuturesApiEndpoint)
        .setBrokerEndpoint(GlobalBrokerApiEndpoint)
        .build();

    const client = new DefaultClient(clientOption);
    const wsService = client.wsService();

    const symbols = await getSpotSymbols(client.restService());
    await spotWsExample(wsService.newSpotPublicWS(), symbols);
    await futuresWsExample(wsService.newFuturesPublicWS());
}

wsReconnectTest().then(() => {
    console.info('Total subscribe: 103');
});

File: sdk/python/test/regression/run_reconnect_test.py
import logging
import os
from time import sleep

from kucoin_universal_sdk.api import DefaultClient, KucoinRestService
from kucoin_universal_sdk.generate.futures.futures_public import FuturesPublicWS
from kucoin_universal_sdk.generate.spot.market import GetAllSymbolsReqBuilder
from kucoin_universal_sdk.generate.spot.spot_public import SpotPublicWS
from kucoin_universal_sdk.model import ClientOptionBuilder, TransportOptionBuilder
from kucoin_universal_sdk.model import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model import WebSocketClientOptionBuilder


def ws_reconnect_test():
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )

    # Retrieve API secret information from environment variables
    key = os.getenv("API_KEY", "")
    secret = os.getenv("API_SECRET", "")
    passphrase = os.getenv("API_PASSPHRASE", "")

    # Set specific options, others will fall back to default values
    ws_client_option = (
        WebSocketClientOptionBuilder().build()
    )

    # Create a client using the specified options
    client_option = (
        ClientOptionBuilder()
        .set_key(key)
        .set_secret(secret)
        .set_passphrase(passphrase)
        .set_websocket_client_option(ws_client_option)
        .set_transport_option(TransportOptionBuilder().build())
        .set_spot_endpoint(GLOBAL_API_ENDPOINT)
        .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
        .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
        .build()
    )
    client = DefaultClient(client_option)
    kucoin_ws_service = client.ws_service()
    symbols = get_spot_symbols(client.rest_service())
    spot_ws_example(kucoin_ws_service.new_spot_public_ws(), symbols)
    futures_ws_example(kucoin_ws_service.new_futures_public_ws())


def get_spot_symbols(rest_service: KucoinRestService):
    market_api = rest_service.get_spot_service().get_market_api()
    symbols_response = market_api.get_all_symbols(GetAllSymbolsReqBuilder().set_market("USDS").build())
    symbols = []
    for d in symbols_response.data:
        symbols.append(d.symbol)
    if len(symbols) > 50:
        symbols = symbols[:50]
    return symbols


def callback_func(topic: str, subject: str, data) -> None:
    pass


def spot_ws_example(spot_public_ws: SpotPublicWS, symbols):
    spot_public_ws.start()

    for symbol in symbols:
        spot_public_ws.trade([symbol], callback_func)

    spot_public_ws.ticker(["BTC-USDT", "ETH-USDT"], callback_func)
    logging.info("Spot subscribe [OK]")


def futures_ws_example(futures_public_ws: FuturesPublicWS):
    futures_public_ws.start()

    futures_public_ws.ticker_v2("XBTUSDTM", callback_func)
    futures_public_ws.ticker_v1("XBTUSDTM", callback_func)
    logging.info("Futures subscribe [OK]")


if __name__ == "__main__":
    ws_reconnect_test()
    logging.info("Total subscribe: 53")
    sleep(3600 * 24 * 365)
